client/my-sites/patterns/components/pattern-gallery/client.tsx

```diff
@@ -14,12 +15,68 @@ import './style.scss';
 
 const LOGGED_OUT_USERS_CAN_COPY_COUNT = 3;
 
+type MasonryGalleryProps = PropsWithChildren< {
+	className?: string;
+	enableMasonry: boolean;
+} >;
+
+// Simulates a Masonry layout by applying negative `margin-top` on every item that doesn't sit in
+// the first row
+function MasonryGallery( { children, className, enableMasonry }: MasonryGalleryProps ) {
+	const ref = useRef< HTMLDivElement >( null );
+
+	useLayoutEffect( () => {
+		if ( ! ref.current || ! enableMasonry ) {
+			return;
+		}
+
+		const element = ref.current;
+
+		function calculateLayout() {
+			const columnCount = getComputedStyle( element ).gridTemplateColumns.split( ' ' ).length;
+
+			if ( columnCount === 1 ) {
+				return;
+			}
+
+			const items = [ ...element.querySelectorAll< HTMLElement >( '.pattern-preview' ) ];
+
+			// We calculate the difference between the top coordinates of each `.pattern-preview`
+			// with the bottom coordinates of the first `.pattern-preview` in the same column. This
+			// value is then used to set a negative `margin-top`, simulating a Masonry layout
+			items.slice( columnCount ).forEach( ( item, i ) => {
+				item.style.marginTop = '0';
+
+				const firstRowBottom = items[ i ].getBoundingClientRect().bottom;
+				const thisRowTop = item.getBoundingClientRect().top;
+
+				item.style.marginTop = `${ firstRowBottom - thisRowTop }px`;
+			} );
+		}
+
+		calculateLayout();
+		element.addEventListener( 'patternPreviewResize', calculateLayout );
```

Comment:
Ideally, I'd prefer not introducing a custom event, if possible. It may be suitable if we're doing something really global, but in this isolated scenario, it might make more sense to rely on context or component state and stay as reactive as possible.

Reply:
Changing this would require a non-trivial amount of refactoring, and we are pressed for time. I would prefer to keep this as-is.